### DISCUSSION TITLE
NextJS v9.3 performance of getServerSideProps vs getInitialProps

### DISCUSSION THREAD

**Discussion by willemliu** [vercel/next.js#10930] [Help]
Hi,

So I've created a little demo to show-case all the different ways NextJS can do page-rendering: [csr, ssr, ssg].
You can find that demo here: https://static-next.willemliu.now.sh/
Code here: https://github.com/willemliu/static-next

In the NextJS documentation the following recommendation can be found:
![image](https://user-images.githubusercontent.com/974906/76304751-332e3600-62c4-11ea-8b85-881134dd967a.png)

However I've found a significant performance penalty in using `getServerSideProps` vs `getInitialProps`

![image](https://user-images.githubusercontent.com/974906/76304863-653f9800-62c4-11ea-912f-be2670c0be19.png)

You can check for yourself by navigating between `SSR` and `Hybrid-SSR` in the demo mentioned above. `SSR` implements `getServerSideProps` and `Hybrid-SSR` implements `getInitialProps`.
Both do exactly the same.

```
export async function getServerSideProps(context: any) {
    const res = await fetch(
        "https://static-next.willemliu.now.sh/api/test"
    ).then((res) => res.json());
    console.log("getServerProps", res, context.params, context.query);
    return {
        props: { ...res, name: context.query.slug }
    };
}
```

```
Index.getInitialProps = async (context: any) => {
    const res = await fetch(
        "https://static-next.willemliu.now.sh/api/test"
    ).then((res) => res.json());
    console.log("getInitialProps", res, context.params, context.query);
    return { ...res, name: context.query.slug };
};
```

It seems to me that the penalty comes from the extra roundtrip to the lambda `getServerSideProps` of that page upon navigation. Where with `getInitialProps` the client-side code runs `getInitialProps` and does the data fetching directly.

Knowing this, my question is what are the benefits of implementing `getServerSideProps` vs `getInitialProps`?

**Comment by timneutkens** [comment]
Covered quite a bit in https://github.com/zeit/next.js/discussions/10920#discussioncomment-1088

- Caching data becomes significantly simpler
- Doesn't affect bundle size, one example is that we currently see people shipping a markdown parser + compiler to the browser instead of pre-processing the markdown. Another is that we see gigantic data fetching libraries being used.
- Can directly run server-side code, for example right now you're fetching an API route in getServerSideProps, but you could import the function behind it and execute it in getServerSideProps if you're using API routes.
- Significantly easier to authorize requests based on cookies

In many of the cases where we see Next.js used there has to be some kind of caching layer as the data is not "live", eg it comes from a CMS that changes less than once per 1 second. 

At [ZEIT](https://zeit.co) the full dashboard is static. If a page does require live data we client-side render it using: https://swr.now.sh/.

**Reply by willemliu** [reply]
> ...
> * Can directly run server-side code, for example right now you're fetching an API route in getServerSideProps, but you could import the function behind it and execute it in getServerSideProps if you're using API routes.
> ...

Just to clarify why the demo uses an API route. In a real-project we would probably consume some endpoint from our service-layer. Hence the API route here to mimic similar behaviour.

**Reply by timneutkens** [reply]
Right, that's just one of the benefits though. Generally the caching bit makes the most sense.

**Reply by Timer** [reply]
Adding my 2c: Even if you're directly calling another API, the effect should be much less pronounced if your deployed app is colocated with your API/data storage (i.e. same region—same data center).

I'd love to reevaluate this with more real-world use case.

**Reply by willemliu** [reply]
> Adding my 2c: Even if you're directly calling another API, the effect should be much less pronounced if your deployed app is colocated with your API/data storage (i.e. same region—same data center).
> 
> I'd love to reevaluate this with more real-world use case.

In case of the example everything runs under the same application. The API route returns static data. So performance should be near optimal. For real projects I can only see a growing latency as real data-sources need to be called upon and maybe some application logic need to be applied.

**Reply by Timer** [reply]
In situations where everything is in the same application, `getServerSideProps` should **never** run a fetch request. It should directly invoke the method that returns the said static data.

**Reply by Timer** [reply]
In "real data-sources", `getServerSideProps` w/ a proper deployment strategy should result in less overall latency (or very similar with only one API call) because you're making one request over the WAN, and all other requests to sub-apis within the same data-center (especially for waterfall data requirements).

**Comment by willemliu** [comment]
> In situations where everything is in the same application, `getServerSideProps` should **never** run a fetch request. It should directly invoke the method that returns the said static data.

I know but the **example** is written this way to mimic api calls. To closely match the architecture we have: REST api's. 
And then we get to the discovery of the original post. That getServerSideProps is slower than getInitialProps.

**Reply by Timer** [reply]
The example is flawed in that it is showcasing a behavior which should never be implemented in a production application.

If you'd like to measure the performance impact of this, please invoke a [3rd party API instead of same-application](https://github.com/zeit/next.js/discussions/10930#discussioncomment-1117) and find the median latency.

This is the proper approach for your example:
```diff
diff --git a/pages/api/test.ts b/pages/api/test.ts
index 7ecb211..28d9e7e 100644
--- a/pages/api/test.ts
+++ b/pages/api/test.ts
@@ -1,15 +1,17 @@
 import Cors from "micro-cors";
 const cors = Cors({ allowMethods: ["GET", "HEAD"] });
 
+export function getData() {
+    return {
+        name: "Home",
+        date: new Date().toLocaleString()
+    };
+}
+
 function test(req, res) {
     res.statusCode = 200;
     res.setHeader("Content-Type", "application/json");
-    res.end(
-        JSON.stringify({
-            name: "Home",
-            date: new Date().toLocaleString()
-        })
-    );
+    res.end(JSON.stringify(getData()));
 }
 
 export default cors(test);
diff --git a/pages/ssr/[slug].tsx b/pages/ssr/[slug].tsx
index d2484e2..ece5db9 100644
--- a/pages/ssr/[slug].tsx
+++ b/pages/ssr/[slug].tsx
@@ -1,5 +1,6 @@
 import "isomorphic-unfetch";
 import { Regenerate } from "../../src/components/Regenerate";
+import { getData } from "../api/test";
 
 function Index(props: any) {
     return (
@@ -35,9 +36,7 @@ function Index(props: any) {
 }
 
 export async function getServerSideProps(context: any) {
-    const res = await fetch(
-        "https://static-next.willemliu.now.sh/api/test"
-    ).then((res) => res.json());
+    const res = await getData();
     console.log("getServerProps", res, context.params, context.query);
     return {
         props: { ...res, name: context.query.slug }
```

**Reply by Timer** [reply]
To be clear: I'm not disagreeing that in your example `getServerSideProps` is slower than `getInitialProps`. It's just improper use (as is)!

**Comment by willemliu** [comment]
> The example is flawed in that it is showcasing a behavior which should never be implemented in a production application.
> 
> If you'd like to measure the performance impact of this, please invoke a [3rd party API instead of same-application](https://github.com/zeit/next.js/discussions/10930#discussioncomment-1117).
> 
> This is the proper approach for your example:
> 
> ```diff
> diff --git a/pages/api/test.ts b/pages/api/test.ts
> index 7ecb211..28d9e7e 100644
> --- a/pages/api/test.ts
> +++ b/pages/api/test.ts
> @@ -1,15 +1,17 @@
>  import Cors from "micro-cors";
>  const cors = Cors({ allowMethods: ["GET", "HEAD"] });
>  
> +export function getData() {
> +    return {
> +        name: "Home",
> +        date: new Date().toLocaleString()
> +    };
> +}
> +
>  function test(req, res) {
>      res.statusCode = 200;
>      res.setHeader("Content-Type", "application/json");
> -    res.end(
> -        JSON.stringify({
> -            name: "Home",
> -            date: new Date().toLocaleString()
> -        })
> -    );
> +    res.end(JSON.stringify(getData()));
>  }
>  
>  export default cors(test);
> diff --git a/pages/ssr/[slug].tsx b/pages/ssr/[slug].tsx
> index d2484e2..ece5db9 100644
> --- a/pages/ssr/[slug].tsx
> +++ b/pages/ssr/[slug].tsx
> @@ -1,5 +1,6 @@
>  import "isomorphic-unfetch";
>  import { Regenerate } from "../../src/components/Regenerate";
> +import { getData } from "../api/test";
>  
>  function Index(props: any) {
>      return (
> @@ -35,9 +36,7 @@ function Index(props: any) {
>  }
>  
>  export async function getServerSideProps(context: any) {
> -    const res = await fetch(
> -        "https://static-next.willemliu.now.sh/api/test"
> -    ).then((res) => res.json());
> +    const res = await getData();
>      console.log("getServerProps", res, context.params, context.query);
>      return {
>          props: { ...res, name: context.query.slug }
> ```

Thanks I'll try to modify the example to hook up with a real external API to check the real-world impact.

**Reply by Timer** [reply]
It'd also be valuable to measure the difference when multiple API calls are made in sequence.

**Reply by willemliu** [reply]
So I've modified the `demo` to make 2 separate external API calls. The call goes to a 3rd party GraphQL endpoint. This is the result:
![image](https://user-images.githubusercontent.com/974906/76357306-bd53ba00-6317-11ea-8c19-e778b876a292.png)
The first two GraphQL calls are made upon visiting `Hybrid-SSR`.
The `server-side-rendered2.json` is made upon visiting `SSR 2`.
The second two GraphQL calls are made upon visiting `Hybrid-SSR 2`.
The `server-side-rendered.json` GraphQL calls are made upon visiting `SSR`.

So there is still a quite difference in `ms` between the `getServerSideProps` and  `getInitialProps` implementation in favor of the latter.

The difference is quite consistent:
![image](https://user-images.githubusercontent.com/974906/76357775-9649b800-6318-11ea-8a6a-7899691419bc.png)

I'm not sure if adding more API calls will eventually turn it in favor of `getServerSideProps`.

The calls are made in parallel though. So I'm going to add another call which is made in sequence.

**Reply by willemliu** [reply]
Added an extra call in sequence:
![image](https://user-images.githubusercontent.com/974906/76358489-cfcef300-6319-11ea-9c46-64efdf249993.png)
screenshot of another try:
![image](https://user-images.githubusercontent.com/974906/76358702-289e8b80-631a-11ea-83c7-2d7e2317237a.png)
Still a little gap there.
So I'll try to add another one in sequence.

**Reply by willemliu** [reply]
This is my final try:
![image](https://user-images.githubusercontent.com/974906/76359101-ddd14380-631a-11ea-9061-70cd9d6afb86.png)
Even with `4` 3rd party API calls where `2` of them are in parallel and `3` are in sequence it seems that `getInitialProps` seems to perform better.

**Reply by Timer** [reply]
Wow, thanks for the exhaustive test cases! I'm using your website from the US and the `getServerSideProps` version is consistently faster for me (100-300ms faster), but this is likely due to my geographical location being further from the MovieSom API.

Could you try deploying to `bru1` instead of `dub1` and see if that reduces the `getServerSideProps` time further? We should be able to see `getServerSideProps` at least match the performance within <100ms considering your region may be closer to the data source.

**Reply by Timer** [reply]
Also, for the `getServerSideProps` version, could you please pass a [server version](https://github.com/zeit/fetch) of [`fetch` to the `ApolloClient ` configuration](https://www.apollographql.com/docs/react/get-started/#configuration-options) to see if it helps?

I'd love to know if `@zeit/fetch` improves the performance on the server:
https://github.com/zeit/fetch

**Reply by willemliu** [reply]
I'll give that a try and will give an update when I have the results.

**Reply by Timer** [reply]
These findings are going to be really valuable. Thanks for taking the time!

**Reply by willemliu** [reply]
Sadly don't have time this week. Will start on it beginning of next week.

**Reply by willemliu** [reply]
![image](https://user-images.githubusercontent.com/974906/76738862-ae339880-676b-11ea-8091-a40ad233c1c8.png)
After deploying to `bru1` I get similar numbers.

**Reply by willemliu** [reply]
> Also, for the `getServerSideProps` version, could you please pass a [server version](https://github.com/zeit/fetch) of [`fetch` to the `ApolloClient ` configuration](https://www.apollographql.com/docs/react/get-started/#configuration-options) to see if it helps?
> 
> I'd love to know if `@zeit/fetch` improves the performance on the server:
> https://github.com/zeit/fetch

I've just checked `zeit/fetch` and saw this message at the bottom of the README.md:
![image](https://user-images.githubusercontent.com/974906/76744645-dd023c80-6774-11ea-9059-38d082ca535f.png)

And in my example I'm using isomorphic-unfetch. Which already uses `node-fetch` on the server-side. So I don't think it does anything for the performance of the `fetch` implementation itself on de server-side.

**Reply by willemliu** [reply]
While searching for some docs about ApolloClient it became clear that it uses global fetch or another fetch implementation if passed as option. In the example I'm using `isomorphic-unfetch` which sets `node-fetch` as the global fetch when on server-side. So I don't think I can improve performance there anymore.

**Reply by asim-basraa** [reply]
May be not directly related but please allow me to add my observation about getServerSideProps.
I have just deployed new version of nextjs app using getServerSideProps and only reason was to minimize bugs caused by node-fetch mostly related to socket hangups errors due to getinitialprops limitations that we had to use rest API data layer.
Our website got 70000 links and it's updated 100s times a day with new pages and tags. Which resulted generating Google sitemap pings, which will kickoff Google parser to parse website again. 
It chocked node-fetch as concurrency gets very high at times causing socket hangups. After changing infrastructure so many times we had to move to getServerSideProps approach and take out fetch calls from backend.
Once we got socket hangups only way to recover was restarting nodejs.
Big thank you for releasing getServerSideProps.

Off course other solution was to manage concurrency with caching and request throttling, using proxies etc. But this solution let us access database directly rather than exposing it to out world. And I can see boost in performance in my use case.

**Comment by vishalvijay** [comment]
Hi

In our case, NextJS website is running under **www.example.com** and all our APIs are served separately under, **api.example.com** (with appropriate CORS headers enabled). Also, **api.example.com** has it's own caching strategy enabled.

As new version of NextJS recommend using `getServerSideProps` instead of `getInitialProps`, we considered migrating our existing `getInitialProps` implementation with `getServerSideProps`. But soon we realized that this change will increase our TTL because `getServerSideProps` is adding an unnecessary proxying layer by routing the API call through **www.example.com**  instead of directly calling the **api.example.com** from the browser.

Also, reading multiple issue threads in Github gave us an understanding that, `getInitialProps` is not deprecated, we can still use that.

So my concern is, the use case for `getInitialProps` and `getServerSideProps` is slightly different.  So I don't think `getServerSideProps` is a replacement for `getInitialProps` and NextJS doc is misleading it by saying it is the recommended way.

So my understanding is =>

- Use `getInitialProps` if the backend API service is decoupled from NextJS.
- Use  `getServerSideProps` if the backend data available in NextJS without any extra API roundtrip. Or avoid CORS by proxying it through NextJS backend.

Please correct me if I'm wrong.

**Reply by willemliu** [reply]
I believe there is an extra point to take into consideration and that's: Environment variables.
Before `getServerSideProps` we had `build.env` and `env` or `serverRuntimeConfig` and `publicRuntimeConfig` to consider. This needed to be configured very carefully and correctly otherwise one risks exposing secret keys to the client-side.
Using `getServerSideProps` has the benefit that you no longer need to consider this.

**Reply by avxkim** [reply]
That makes sense, still don't get, why do they recommend using `getServerSideProps` for everyone, they consider that everyone's using next api routes on the same domain?

**Reply by msreekm** [reply]
Another big limitation with getServerSideProps is - if you are using graphql and is using apollo client, apollo client can walk down the component hierarchy using getInitialProps and execute the graphql queries at each level , whereas with getServerSideProps you have to move all your queries to top level page component. Hopefully getInitialProps is not going away soon! if they deprecate getInitialProps ,lot of people will move to gatsby!

**Reply by francisrod01** [reply]
I've moved to Angular 11 and I'm happy about that. :)
